Feature/num 128 virologischer befund

NormanWeinert opened the pull request:
Adds the converter for the "Virologischer Befund" template.

Resolves: https://github.com/ehrbase/num_platform/issues/128

subigre committed:
Implement Provide Consent, Provide Patient and Provide Medication Statement transactions

ppazos committed:
fixes for #212 still need to fix some other robot tests, WIP

ZhuYufei990 committed:
fix error bei PAtienten Aufenthalt

ZhuYufei990 committed:
Merge branch 'feature/NUM_119_Stat_Versorgungsfall+Patientenaufenthalt' of https://github.com/ehrbase/fhir-bridge into feature/NUM_119_Stat_Versorgungsfall+Patientenaufenthalt

ZhuYufei990 committed:
fix code smell

ZhuYufei990 committed:
fix code smells again

ZhuYufei990 committed:
fix switch case

ppazos committed:
all test fixed for #212

SevKohler committed:
Merge pull request #335 from ehrbase/feature/212_heart_rate_tests

Feature/212 heart rate tests

subigre committed:
Implement Provide Procedure and Provide Questionnaire-Response

subigre committed:
Merge branch 'develop' into feature/324_conditional_create

subigre committed:
Implement Provide Diagnostic Report

subigre committed:
Implement Provide Observation

subigre committed:
Fix search issue

subigre committed:
Add integration tests for Condition and update Postman collection

subigre committed:
Add integration tests for Consent and update Postman collection

subigre committed:
Add integration tests for DiagnosticReport and update Postman collection

subigre committed:
Add integration tests for Medication Statement and update Postman collection

subigre committed:
Add integration tests for Observation and update Postman collection

subigre committed:
Add integration tests for Patient and update Postman collection

subigre committed:
Add integration tests for Procedure and update Postman collection

subigre committed:
Add integration tests for QuestionnaireResponse and update Postman collection

ZhuYufei990 committed:
fix code smell & add test cases

ZhuYufei990 committed:
fix time value in test case

subigre committed:
Fix SonarQube issues:
- Dedicated Exception
- Remove duplicate literal

subigre committed:
Merge branch 'develop' into feature/294-update-template-startup

# Conflicts:
#	pom.xml

subigre committed:
Add a configuration property to enable the modification of existing templates

SevKohler committed:
Merge pull request #296 from ehrbase/feature/294-update-template-startup

Feature/294 update template startup

subigre committed:
Fix remaining SonarQube issues + revert 'var' usage

SevKohler committed:
Merge pull request #336 from ehrbase/feature/324_conditional_create

Feature/327 conditional create

ZhuYufei990 committed:
Refactoring: rename packages

virologisherBefund ->geccoVirologischerbefund;
DnrAnordnung -> dnranordnung

SevKohler committed:
Merge pull request #338 from ehrbase/feature/334_rename_packages

Refactoring: rename packages

ZhuYufei990 committed:
Refactoring for review requests

subigre committed:
Resolves issue #337

ZhuYufei990 committed:
Merge branch 'develop' into feature/NUM_119_Stat_Versorgungsfall+Patientenaufenthalt

SevKohler committed:
Merge pull request #339 from ehrbase/feature/337-refactoring-find-transaction

Feature/337 Refactoring Find Transaction

ZhuYufei990 committed:
fix with Refactoring in develop

ZhuYufei990 committed:
Merge branch 'develop' into feature/NUM_119_Stat_Versorgungsfall+Patientenaufenthalt

ZhuYufei990 committed:
fix  EncounterRoutes

SevKohler committed:
Merge pull request #333 from ehrbase/feature/NUM_119_Stat_Versorgungsfall+Patientenaufenthalt

Feature/num 119 stat versorgungsfall+patientenaufenthalt

subigre committed:
Update README.md

subigre committed:
Minor fix according to the Wiki documentation update

subigre committed:
Merge remote-tracking branch 'origin/develop' into develop

subigre committed:
Upgrade Spring Boot and IPF

SevKohler committed:
Merge pull request #343 from ehrbase/feature/342_upgrade_dependencies

Upgrade Spring Boot and IPF

NormanWeinert committed:
Tests added

NormanWeinert committed:
Test update

NormanWeinert committed:
Refactoring

NormanWeinert committed:
Ergebnis-Status mapping update

NormanWeinert committed:
Zugehorige Laborprobe choice check updated

NormanWeinert committed:
Cleanup + fix Virusnachweistest mapping in the case of existent ValueQuantity

SevKohler committed:
regenerated opt

subigre committed:
#345 Create an EHR when it is missing

subigre committed:
Fix BundleValidator

subigre committed:
Fix SonarCloud issues

subigre committed:
Fix issues with 'deprecated' test cases

subigre committed:
Fix issue for Observation

subigre committed:
SonarCloud + Bundle tests using RobotFramework

subigre committed:
Revert back to original validation order

subigre committed:
Minor change based on @SevKohler comment

subigre committed:
Merge pull request #346 from ehrbase/feature/345_create_ehr

#345 Create an EHR when it is missing

subigre committed:
Implements OAuth2 authentication with EHRbase

SevKohler committed:
updated immunization

SevKohler committed:
finished

SevKohler committed:
Merge branch 'develop' into feature_78/mapping_vaccination

# Conflicts:
#	config/postman/fhir-bridge.postman_collection.json
#	src/main/java/org/ehrbase/fhirbridge/config/ConversionConfiguration.java
#	src/main/java/org/ehrbase/fhirbridge/config/FhirJpaConfiguration.java
#	src/main/java/org/ehrbase/fhirbridge/ehr/converter/generic/TimeConverter.java
#	src/main/java/org/ehrbase/fhirbridge/fhir/common/Profile.java
#	src/main/java/org/ehrbase/fhirbridge/fhir/common/audit/FhirBridgeEventType.java
#	src/main/java/org/ehrbase/fhirbridge/fhir/medicationstatement/CreateMedicationStatementProvider.java

SevKohler committed:
finished

subigre committed:
Replace 'Create Immunization' by 'Provide Immunization' using generic Camel processors

SevKohler committed:
Merge remote-tracking branch 'origin/develop' into feature_78/mapping_vaccination

# Conflicts:
#	src/main/java/org/ehrbase/fhirbridge/fhir/support/Resources.java

SevKohler committed:
added merge

SevKohler committed:
Merge pull request #351 from ehrbase/feature_78/mapping_vaccination

Feature 78/mapping vaccination

subigre committed:
Merge Postman Collection and Environment

subigre committed:
Merge branch 'develop' into feature/349_ehrbase_oauth2_authentication

subigre committed:
Merge pull request #350 from ehrbase/feature/349_ehrbase_oauth2_authentication

Implements OAuth2 authentication with EHRbase

subigre committed:
Add IT for using search for patient using identifier

subigre committed:
Update patient process

subigre committed:
Code refactoring regarding subject and patient handling

subigre committed:
Deletes deprecated test as the EHR is now automatically created

subigre committed:
Fixes test data

subigre committed:
Initialize HAPI FHIR Validation API at startup (thanks to Hauke)

subigre committed:
Review Encounter resource implementation

SevKohler committed:
Merge remote-tracking branch 'origin/develop' into fix_152

SevKohler committed:
opt updated

subigre committed:
Fix Bundle implementation

SevKohler committed:
Merge pull request #355 from ehrbase/fix_152

Fix 152

subigre committed:
Fix Blood Gas Panel

SevKohler committed:
fix exurity hotspots

SevKohler committed:
changed log string

SevKohler committed:
renamed smells

SevKohler committed:
moved codes to opt generated package

SevKohler committed:
two more moved to opt

SevKohler committed:
added messages

SevKohler committed:
fixed messages to enum

SevKohler committed:
Merge pull request #358 from ehrbase/fix_#357

renamed Name smells

subigre committed:
Add javadoc and fix issue with identifier

SevKohler committed:
Merge pull request #360 from ehrbase/fix_#359

moved codes to opt generated package

subigre committed:
Merge branch 'develop' into feature/352_patient_improvement

# Conflicts:
#	src/main/java/org/ehrbase/fhirbridge/config/ConversionConfiguration.java
#	src/main/java/org/ehrbase/fhirbridge/fhir/bundle/validator/BloodGasPanelBundleValidator.java

subigre committed:
Fix Sonar issue: CognitiveComplexity

SevKohler committed:
Merge pull request #356 from ehrbase/fix_340

Fix 340

subigre committed:
Bugfix ArrayIndexOutOfBoundsException in PCRObservationConverter

subigre committed:
Merge branch 'develop' into feature/352_patient_improvement

# Conflicts:
#	src/main/java/org/ehrbase/fhirbridge/ehr/converter/generic/ConditionToObservationConverter.java
#	src/main/java/org/ehrbase/fhirbridge/ehr/converter/generic/EncounterToAdminEntryConverter.java

subigre committed:
Remove createIdentifierSection method

subigre committed:
Fix issue with IT linked to previous removed method

SevKohler committed:
refactored Bloodpressure

subigre committed:
Merge pull request #361 from ehrbase/feature/352_patient_improvement

Feature/352 patient improvement

SevKohler committed:
refactored Blood pressure

SevKohler committed:
refactored body height

SevKohler committed:
refactored body height

SevKohler committed:
delete wrong template and reload new one, changed mapping accordingly

SevKohler committed:
Merge pull request #362 from ehrbase/refactor_caridnality_mapping

Refactor caridnality mapping

SevKohler committed:
Merge pull request #363 from ehrbase/refactor_bodyheight

Refactor bodyheight

subigre committed:
Updated release version to v1.2.0

subigre committed:
Merge pull request #365 from ehrbase/release/v1.2.0

Updated release version to v1.2.0

subigre committed:
Merge branch 'release/v1.2.0'

subigre committed:
Merge branch 'master' into develop

SevKohler committed:
refactored bodyweight

SevKohler committed:
Merge remote-tracking branch 'origin/develop' into refactor_bodyweight

SevKohler committed:
refactor clinical frailty

subigre committed:
Fix ITs

SevKohler committed:
fix spelling mistake

SevKohler committed:
Merge pull request #367 from ehrbase/refactor_bodyweight

Refactor bodyweight

SevKohler committed:
refactored diagnostic lab

SevKohler committed:
fixed conversionService

SevKohler committed:
refactor fio2

SevKohler committed:
Merge pull request #370 from ehrbase/fix_spelling_mistake

fix spelling mistake

SevKohler committed:
Merge pull request #374 from ehrbase/refactor_dnr

fixed conversionService

SevKohler committed:
refactoring in progress

SevKohler committed:
Merge pull request #373 from ehrbase/refactor_fio2

Refactor fio2

SevKohler committed:
Merge pull request #372 from ehrbase/refactor_diagnosticLab

Refactor diagnostic lab

subigre committed:
Fix ITs

subigre committed:
Merge pull request #364 from ehrbase/refactore_body_temperature

delete wrong template for Körpertemperatur and reload new one, changed mapping accordingly

subigre committed:
Fix integration tests

SevKohler committed:
added Time converter

SevKohler committed:
Merge pull request #368 from ehrbase/refactor_clinical_frailty

refactor clinical frailty

SevKohler committed:
fixed tests

SevKohler committed:
added new converter

SevKohler committed:
deleted the generic converter

SevKohler committed:
refactored heart rate

SevKohler committed:
started developing tests

SevKohler committed:
fixed smells

SevKohler committed:
tests added

SevKohler committed:
deleted heartrate test from observation

SevKohler committed:
refactored history of Travel

SevKohler committed:
refactored history of Travel

subigre committed:
Refactor symptom

subigre committed:
Merge branch 'refactor_gecco_diagnose' into refactoring/symptom

SevKohler committed:
fixed tests

SevKohler committed:
changed Exceptions

SevKohler committed:
fixed tests

SevKohler committed:
application.yml debug set to false

SevKohler committed:
Merge pull request #375 from ehrbase/refactor_gecco_diagnose

Refactor gecco diagnose

SevKohler committed:
Merge pull request #377 from ehrbase/refactor_heartrate

refactor heartrate

SevKohler committed:
Merge pull request #378 from ehrbase/refactor_history_of_travel

Refactor history of travel

SevKohler committed:
changed mispellings

SevKohler committed:
refactoring started

SevKohler committed:
Merge remote-tracking branch 'origin/develop' into refactor_impfstatus

# Conflicts:
#	src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/heartrate/HerzfrequenzObservationConverter.java
#	src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/historyoftravel/HistoryOfTravelCompositionConverter.java
#	src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/historyoftravel/ReisehistorieAdminEntryConverter.java

SevKohler committed:
merged with develop

SevKohler committed:
fixed merge

subigre committed:
Refactor symptom

SevKohler committed:
fixed tests

SevKohler committed:
finished refactoring sars cov

SevKohler committed:
finished refactoring sars cov

SevKohler committed:
Merge remote-tracking branch 'origin/develop' into refactor_known_exposure

# Conflicts:
#	src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/knownexposure/SarsCov2ExpositionEvaluationConverter.java

SevKohler committed:
Merge pull request #379 from ehrbase/refactor_impfstatus

Refactor impfstatus

SevKohler committed:
refactor medication

SevKohler committed:
refactor medication

SevKohler committed:
Merge pull request #380 from ehrbase/refactor_known_exposure

Refactor known exposure

SevKohler committed:
Merge pull request #381 from ehrbase/refactor_medication

Refactor medication

subigre committed:
Refactor therapy converter

subigre committed:
Fix robot tests

SevKohler committed:
Merge pull request #383 from ehrbase/refactoring/symptom

Refactoring/symptom

subigre committed:
Merge branch 'develop' into refactor/therapy

# Conflicts:
#	src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/therapy/NichtDurchgefuehrteProzedurEvaluationConverter.java
#	src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/therapy/ProzedurActionConverter.java
#	src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/therapy/TherapyCompositionConverter.java
#	src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/therapy/UnbekannteProzedurEvaluationConverter.java

SevKohler committed:
started refactoring

SevKohler committed:
speciman converter refactoring

subigre committed:
Merge pull request #382 from ehrbase/refactor/therapy

Refactor therapy converter

SevKohler committed:
finished refactoring speciman converter

SevKohler committed:
refactor observation lab

subigre committed:
Refactor SOFA score

SevKohler committed:
bug fix implemented

SevKohler committed:
updated tests

subigre committed:
Enables Spring Boot Actuator endpoints: Loggers, LogFile and Liquibase

SevKohler committed:
Merge pull request #386 from ehrbase/refactor_gecco_diagnose

Refactor gecco diagnose

subigre committed:
Fix robot tests after updating SOFA score profile

subigre committed:
Merge pull request #387 from ehrbase/feature/384_spring_boot_actuator

Enables Spring Boot Actuator endpoints: Loggers, LogFile and Liquibase

subigre committed:
Fix robot tests after updating SOFA score profile

subigre committed:
Fix robot tests after updating SOFA score profile

subigre committed:
Fix robot tests after updating SOFA score profile

subigre committed:
Merge branch 'develop' into refactoring/sofascore

# Conflicts:
#	tests/robot/OBSERVATION/10_create_sofa_score.robot

subigre committed:
Merge pull request #385 from ehrbase/refactoring/sofascore

Refactor SOFA score

subigre committed:
Updated release version to v1.2.1

subigre committed:
Merge pull request #388 from ehrbase/release/v1.2.1

Updated release version to v1.2.1

subigre committed:
Bump version to 1.3.0-SNAPSHOT

SevKohler committed:
refactoring finished

SevKohler committed:
fixed tests

SevKohler committed:
refactored smells

SevKohler committed:
Deleted return

NormanWeinert committed:
Ergebnis-Status fix + Refactor with current SDK Generator version

SevKohler committed:
Merge pull request #389 from ehrbase/refactor_observationLab

Refactor observation lab

SevKohler committed:
refactored

subigre committed:
Fix tests:
- 087 Create Body Temperature loinc period
- 088 Create Body Temperature loinc period_2
- 090 Create Body Temperature snomed period
- 091 Create Body Temperature snomed period_2

SevKohler committed:
refactor patient

SevKohler committed:
added todo

subigre committed:
Fix tests:
- 079 Create Condition Symptoms-Covid-19 icd10
- 080 Create Condition Symptoms-Covid-19 alphaid

NormanWeinert committed:
Merge branch 'develop' into feature/NUM-128_Virologischer-Befund

subigre committed:
Fixed json files and update invalid test cases

wlad committed:
downgrade CircleCI resource classe

wlad committed:
Merge pull request #397 from ehrbase/feature/CI-downgrade-resource-class

downgrade CircleCI resource classe

SevKohler committed:
refactor patient

SevKohler committed:
Merge pull request #398 from ehrbase/refactor_diagnose

refactored diagnose

SevKohler commented on src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/virologischerbefund/BefundJedesEreignisPointEventConverter.java:
```diff
@@ -0,0 +1,110 @@
+        }
+    }
+
+    private void mapProbe(BefundJedesEreignisPointEvent befundevent, Specimen specimen) throws FHIRException {
```
Method too long please split in smaller methods

NormanWeinert replied:
Simplifier Link: https://simplifier.net/medizininformatikinitiative-highmed-ic/virologischerbefund
Highmed Link: http://88.198.146.13/ckm/templates/1246.169.636

I used the canonical link (http://highmed.org/StructureDefinition/VirologischerBefund) in the Structure Definition of the simplifier site, but apparently it is still in draft and can still change. It hasn't updated to a new one on the simplifier site yet.

NormanWeinert replied:
As I just saw that the canonical links for the "Diagnostic Report" and "Specimen"  also changed I will post them here as well. 
Diagnostic Report: https://simplifier.net/coreprofilesstu3/diagnosticreport
Specimen: https://simplifier.net/coreprofilesstu3/specimen

SevKohler replied:
Okay, before i start to review, can you ensure again that the cardinality is always checked if the field in fhir is not a must have ? Since this is very often a bug people tend to do, would do me a great favour, since we are currently super busy. 
Thanks.
I will start the review as soon as i have time (as you see currently we are doing a great deal of stuff at the bridge specifically for gecco).

NormanWeinert replied:
I don't think I have an extra cardinality check for every optional fhir rescource and it would just throw a standard error when trying to access it. 
As a result, I will go through the code again and implement a "if not present throw exception" check on every used optional fhir rescource when it shows up.
I can also combine this with updating the outdated links and splitting up that method you already mentioned. Just need to ask in a meeting tomorrow which links I can reliably use. I probably should be able to commit and push the adjustments tomorrow.

SevKohler replied:
I would suggest, if the field ist not necessary in openEHR, you dont map it (use return Optional.empty()) 
Example: 
https://github.com/ehrbase/fhir-bridge/blob/c1aec64a9e1f677330d247720674ebeb44c346c3/src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/bodyheight/GroesseLaengeObservationConverter.java#L19

All classes until Patient, have been refactored to do that (diagnose is an exception since this one is only not deleted in case we need it later on), so you can take a look how its done there. The last missing classes (everything beyond patient to smokingStatus are in the making).

If the code/value is not valid you need to throw an:
ConversionException. 

This can be sometimes a pain in the arse (Especially when talking about Codings and CodableConcept since this requires iteration), i know, but it has to be done, otherwise the bridge will crash with an normal exception, this is not what we want.  Remember if the field is REQUIRED in fhir you dont need to do this check, since HAPI does it :)

Thanks

SevKohler reviewed (commented):
I need the virologischer befund e.g. simplifier link in order to make a review. The one of the structure definition is not working.

wlad committed:
Merge pull request #391 from ehrbase/bugfix/217_unexpected_errors

Enable tests with updated profile

wlad committed:
Merge pull request #393 from ehrbase/bugfix/190_errors_alternative_paths

Update tests according to the 'Symptoms Covid-19' profile

SevKohler committed:
refactoring continues

wlad committed:
Merge pull request #395 from ehrbase/bugfix/194_unexpected_errors

Fixed json files and update invalid test cases

SevKohler committed:
finished mapping

SevKohler committed:
finished first refactoring, tests and template need to be updated with the template update in the future

SevKohler committed:
deleted useless tests

SevKohler committed:
finished refactoring

SevKohler committed:
correct tests

SevKohler committed:
Merge pull request #400 from ehrbase/refactor_patient

Refactor patient

SevKohler committed:
Merge pull request #401 from ehrbase/refactor_patient_discharge

Refactor patient discharge

NormanWeinert committed:
mapProbe method splt + Profile links updated + exceptions updated + cardinality checks for resources

NormanWeinert commented:
@SevKohler 
- I've split the method you marked into smaller ones.

- The Profile links have been updated:
Virologischer Befund: https://simplifier.net/medizininformatikinitiative-highmed-ic/virologischerbefund
Diagnostic Report: https://simplifier.net/medizininformatikinitiative-highmed-ic/highmedicdiagnosticreport
Specimen: https://simplifier.net/medizininformatikinitiative-highmed-ic/specimen-duplicate-5
It still could be the case that they need to be updated later if the canonical links finally become stable.

- I've updated the exceptions to ConversionExceptions and added cardinality checks.
  Every optional fhir resource should now be checked if it is indeed present and if not throw a suitable ConversionException.
  As you mentioned Codings were a big part as I also needed to check not only if they are present but also that there are not too 
  many present.
  I only have 2 optional open EHR resources:
  One is bound to a required resource. Either ValueCodeableConcept or ValueQuantity needs to be present and only in the 
  ValueQuantity case the optional value is mapped together with a fixed value for the required one. So there was already a check 
  in place.
  The other one will simply not be mapped if the required values are not present. Still, exceptions will be thrown if one of the 
  optional values is present in an incomplete state.

I hope I've catched everything in that regard so that your job is a bit easier. If I missed something please do not hesitate to inform me.

----------------------------------------------------------------------------------------------------------------------------------------

Still, one more issue of the Virologischer Befund mapping was found through testing though, but as I observed the mapping should be correct and the pull request should continue as implementing the conversion is currently high priority here:

Comparing the fhir input from "create-virologischer-befund.json" to the result in the ehrbase in "paragon-create-virologischer-befund.json" still throws out a difference:

ObjectRemoved{ object removed: org.ehrbase.fhirbridge.ehr.opt.virologischerbefundcomposition.VirologischerBefundComposition/#befund/jedesEreignis/0/labortestPanel/proAnalyt/0/ergebnisStatus/0/value2/1 }
ListChange{ 'befund/jedesEreignis/0/labortestPanel/proAnalyt/0/ergebnisStatus/0.value2' collection changes :
  1. 'org.ehrbase.fhirbridge.ehr.opt.virologischerbefundcomposition.VirologischerBefundComposition/#befund/jedesEreignis/0/labortestPanel/proAnalyt/0/ergebnisStatus/0/value2/1' removed }

This means that there is a value in the ehrbase that is not present on the fhir side of the conversion. Again, the value seems to be an "Ergebnis-Status" as in the previous issue. I checked my code and the "Ergebnis-Status" value is only added once and in the correct state.

I also checked both the fhir and ehrbase mapping and on every side all nessesary values are present and there are also no extra/duplicated values.

So this seems to me like another bug outside of my conversion.

Maybe it has something to do with the validator again, but I am not sure.

Still, as the result of the conversion seems to be correct in my eyes the code should still be pulled and this issue could be opened seperately and fixed afterwards.

I don't know where it is located though. I could open it here on the fhir-bridge, but it could end up being just another variation of the ehrbaseSDK validator issue. What would be your advice here?

subigre committed:
Fixes NullPointerException in ProvideResourceAuditHandler + additional audit info

subigre committed:
Fixes NullPointerException in ProvideResourceAuditHandler + additional audit info

subigre committed:
Fixes NullPointerException in ProvideResourceAuditHandler + additional audit info

subigre committed:
Fixes NullPointerException in ProvideResourceAuditHandler + additional audit info

subigre committed:
Fixes NullPointerException in ProvideResourceAuditHandler + additional audit info

subigre committed:
Merge remote-tracking branch 'origin/bugfix/402_nullpointerexception_audit' into bugfix/402_nullpointerexception_audit

subigre committed:
Merge branch 'develop' into bugfix/402_nullpointerexception_audit

subigre committed:
Merge pull request #405 from ehrbase/bugfix/402_nullpointerexception_audit

Bugfix/NullpointerException audit

subigre committed:
Re-apply patch

subigre committed:
Re-apply patch

subigre committed:
Re-apply patch after rebase

subigre committed:
Reapply changes after rebase from develop

subigre committed:
Reapply changes after rebase from develop

NormanWeinert committed:
Revert "Merge branch 'develop' into feature/NUM-128_Virologischer-Befund"

This reverts commit 906516354e230e3b4760504bc6986c9c4a0f20bb, reversing
changes made to 9bee8fa8573ceae590c4a5b6c562040a7f595877.

SevKohler commented:
@NormanWeinert dont forget to fix the Code smells except "Remove this useless assignment to local variable" that one you can ignore.

NormanWeinert committed:
Address code smells

NormanWeinert commented:
@SevKohler 
Code Smells that are not "Remove this useless assignment to local variable" have been addressed in the latest commit with 1 exception:

In "src/.../fhirbridge/fhir/support/Bundles.java":
"Refactor this method to reduce its Cognitive Complexity from 6 to the 5 allowed."

public static Profile getTransactionProfile(Bundle bundle) {
        for (Bundle.BundleEntryComponent entry : bundle.getEntry()) {
            if (Resources.hasProfile(entry.getResource(), Profile.BLOOD_GAS_PANEL)) {
                return Profile.BLOOD_GAS_PANEL;
            } else if (Resources.hasProfile(entry.getResource(), Profile.ANTI_BODY_PANEL)) {
                return Profile.ANTI_BODY_PANEL;
            } else if (Resources.hasProfile(entry.getResource(), Profile.DIAGNOSTIC_REPORT_LAB)) {
                return Profile.DIAGNOSTIC_REPORT_LAB;
            } else if (Resources.hasProfile(entry.getResource(), Profile.VIROLOGISCHER_BEFUND)) {
                return Profile.VIROLOGISCHER_BEFUND;
            }
        }
        return null;
    }

I only added the VIROLOGISCHER_BEFUND section. 
Whoever created this class would need to refactor it so that not every new bundle increases the Cognitive Complexity as currently every new entry is bound to increase it.

subigre committed:
Comments gh-pages deployment

SevKohler commented:
@NormanWeinert put the if in a different method and there you go. Also never return null. If null is needed please use Optional.empty(); . This does not create exceptions etc. .

NormanWeinert committed:
Refactored Bundles.java

NormanWeinert commented:
@SevKohler 
I refactored Bundles.java and used optional. The smell should be gone.
My tests say that everything is in order, but again as I did not create this class and do not know all workings behind the scenes I cannot guarantee 100% correctness.

NormanWeinert commented:
@SevKohler  Ok, now the smell is truely gone and all tests have passed.

SevKohler commented on src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/virologischerbefund/BefundJedesEreignisPointEventConverter.java:
```diff
@@ -0,0 +1,122 @@
+
+    private void mapLabortestbezeichnung(Observation observation, BefundJedesEreignisPointEvent befundevent) {
+
+        if(observation.getCategory().get(0).getCoding().get(2).getCode().equals("122442008")) {
```
Please iterate ALL codings (since a JSON is not sorted) and check for the code. 
If not code found -> throw.

NormanWeinert replied:
All codings are now checked with a new boolean function with a for each function. Only if the correct code is found somewhere the fixed value will be mapped.
https://github.com/ehrbase/fhir-bridge/pull/392/files#diff-7e8996e1193679795973b45c1d8e9b897e57859669d2d6a01138153dfe06ad49

SevKohler commented on src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/virologischerbefund/BefundJedesEreignisPointEventConverter.java:
```diff
@@ -0,0 +1,122 @@
+
+        ProbeCluster probecluster = new ProbeCluster();
+
+        if (!specimen.hasCollection()){
```
A specimen does not NEED to have a collections in FHIR (see profile).
Please check if it exists -> map it. 
If it does not exist -> dont map it. 
Use Optional for that, like in the example i showed you.

NormanWeinert replied:
Check has been removed and Optional is used.
https://github.com/ehrbase/fhir-bridge/pull/392/files#diff-7e8996e1193679795973b45c1d8e9b897e57859669d2d6a01138153dfe06ad49

SevKohler commented on src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/virologischerbefund/BefundJedesEreignisPointEventConverter.java:
```diff
@@ -0,0 +1,122 @@
+        if (!specimen.hasCollection()){
+            throw new ConversionException("Specimen resource needs to have a Collection.");
+        }
+        mapZeitpunktDerProbenentnahme(probecluster, specimen);
```
This should be in the TimeConverter. There is even a method for that, as far as i can see it. 
convertSpecimanCollection

NormanWeinert replied:
Now convertSpecimanCollection from the TimeConverter is called which is included in an extra method that returns Optional<TemporalAccessor>.
https://github.com/ehrbase/fhir-bridge/pull/392/files#diff-7e8996e1193679795973b45c1d8e9b897e57859669d2d6a01138153dfe06ad49

SevKohler commented on src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/virologischerbefund/BefundJedesEreignisPointEventConverter.java:
```diff
@@ -0,0 +1,122 @@
+    private void mapNameDerKoerperstelle(ProbeCluster probecluster, Specimen specimen) throws FHIRException {
+
+        AnatomischeLokalisationCluster anatomischeLokalisationCluster = new AnatomischeLokalisationCluster();
+        if (!specimen.getCollection().getBodySite().hasCoding() || specimen.getCollection().getBodySite().getCoding().size() != 1){
```
Do not check for size. 
BodySite is mandatory 1..1, if you leave enter NO coding HAPI will throw an exception. 
So we do not need to check that.

NormanWeinert replied:
Checks have been removed.
https://github.com/ehrbase/fhir-bridge/pull/392/files#diff-7e8996e1193679795973b45c1d8e9b897e57859669d2d6a01138153dfe06ad49

SevKohler commented on src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/virologischerbefund/BefundJedesEreignisPointEventConverter.java:
```diff
@@ -0,0 +1,122 @@
+        if (!specimen.getCollection().getBodySite().hasCoding() || specimen.getCollection().getBodySite().getCoding().size() != 1){
+            throw new ConversionException("Body Site of Specimen resource needs to have exactly one instance of Coding.");
+        }
+        switch (specimen.getCollection().getBodySite().getCoding().get(0).getCode()) {
```
I would maybe use a for each here, does the template allow multiple values ? 
If not, for each and map the first one, or request an alteration of the template if it is logical to ahve several body sites here (which i do not think :D )

NormanWeinert replied:
For each is now used. Also, getCodesAsMap replaces the switch checks.
https://github.com/ehrbase/fhir-bridge/pull/392/files#diff-7e8996e1193679795973b45c1d8e9b897e57859669d2d6a01138153dfe06ad49

SevKohler commented on src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/virologischerbefund/ProAnalytClusterConverter.java:
```diff
@@ -0,0 +1,93 @@
+
+        mapVirusnachweistest(observation, proAnalytCluster);
+
+        if (observation.hasValueCodeableConcept()){
```
Put into own method pls.

NormanWeinert replied:
It has its own method mapValue now.
https://github.com/ehrbase/fhir-bridge/pull/392/files#diff-4d8c593b51ad51f7cbd0a2addd7fb06a59da59dc9c48d4cf9a9e1edb3fa9dc5e

SevKohler commented on src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/virologischerbefund/ProAnalytClusterConverter.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    private void mapVirusnachweistest(Observation observation, ProAnalytCluster proAnalytCluster){
+        if (!observation.getCode().hasCoding() || observation.getCode().getCoding().size() != 1){
```
also does not require check ssince its 1..1. 
Nevertheless it could be several Codings so use for each as explained above.

NormanWeinert replied:
For each is now being used.
https://github.com/ehrbase/fhir-bridge/pull/392/files#diff-4d8c593b51ad51f7cbd0a2addd7fb06a59da59dc9c48d4cf9a9e1edb3fa9dc5e

SevKohler commented on src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/virologischerbefund/ProAnalytClusterConverter.java:
```diff
@@ -0,0 +1,93 @@
+            throw new ConversionException("Code of Observation needs to have exactly one instance of Coding.");
+        }
+        switch (observation.getCode().getCoding().get(0).getCode()){
+            case "94500-6":
```
Avoid strings, does the DefiningCode contain thsi Code ? If yes please use the Enum to compare. 
Also this can then be done by adding an getCodesAsMap method wher you just use the keys to map, so you dont need switch case. Check also the other switch cases for that.

NormanWeinert replied:
Strings have been removed and getCodesAsMap is used in getVirusCode instead.
https://github.com/ehrbase/fhir-bridge/pull/392/files#diff-4d8c593b51ad51f7cbd0a2addd7fb06a59da59dc9c48d4cf9a9e1edb3fa9dc5e

SevKohler commented on src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/virologischerbefund/ProAnalytClusterConverter.java:
```diff
@@ -0,0 +1,93 @@
+            throw new ConversionException("ValueCodeableConcept of Observation needs to have exactly one instance of Coding.");
+        }
+        switch(observation.getValueCodeableConcept().getCoding().get(0).getCode()){
+            case "260373001":
```
getCodeAsMap if possible

NormanWeinert replied:
GetCodeAsMap is now being used in getVirusCode.
https://github.com/ehrbase/fhir-bridge/pull/392/files#diff-4d8c593b51ad51f7cbd0a2addd7fb06a59da59dc9c48d4cf9a9e1edb3fa9dc5e

SevKohler commented on src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/virologischerbefund/ProAnalytClusterConverter.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    private void mapNachweistest(Observation observation, ProAnalytCluster proAnalytCluster) throws FHIRException {
+        if (observation.getValueCodeableConcept().getCoding().size() != 1){
```
no check required

NormanWeinert replied:
Check has been removed.
https://github.com/ehrbase/fhir-bridge/pull/392/files#diff-4d8c593b51ad51f7cbd0a2addd7fb06a59da59dc9c48d4cf9a9e1edb3fa9dc5e

SevKohler commented on src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/virologischerbefund/ProAnalytClusterConverter.java:
```diff
@@ -0,0 +1,93 @@
+        /**
+         * ZugehoerigeLaborprobe is optional in openEHR and will only be mapped if one of the required resources is present.
+         */
+        if (observation.getSpecimen().hasIdentifier()){
```
Own method

NormanWeinert replied:
Is now in its own method as an Optional.
https://github.com/ehrbase/fhir-bridge/pull/392/files#diff-4d8c593b51ad51f7cbd0a2addd7fb06a59da59dc9c48d4cf9a9e1edb3fa9dc5e

SevKohler commented on src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/virologischerbefund/VirologischerBefundCompositionConverter.java:
```diff
@@ -0,0 +1,51 @@
+
+    private void mapFallIdentifikation(VirologischerBefundComposition composition, DiagnosticReport diagnosticReport) {
+        FallidentifikationCluster fallidentifikationCluster = new FallidentifikationCluster();
+        if (!diagnosticReport.hasEncounter()){
```
As far as i can see in the Profile this is not required, therefore if not existent do not map it. 
Also is Encounter the right field to be mapped for Fallidentifkation, read the descriptions for both. 
Im not sure, your the mapper just double check if its correct.

NormanWeinert replied:
Is now an Optional.
DiagnosticReport.encounter is the correct value to map for Fallidentifikation for this Template after my table. 
https://github.com/ehrbase/fhir-bridge/pull/392/files#diff-7946f6e29b883aeca8c67cd3043f2779692897e990650ff6d5216a3b6b27e0a6

SevKohler commented on src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/virologischerbefund/VirologischerBefundCompositionConverter.java:
```diff
@@ -0,0 +1,51 @@
+        if (!diagnosticReport.hasEncounter()){
+            throw new ConversionException("Diagnostic Report needs to have an Encounter.");
+        }
+        if (!diagnosticReport.getEncounter().hasReference()){
```
Same

NormanWeinert replied:
Check removed.
https://github.com/ehrbase/fhir-bridge/pull/392/files#diff-7946f6e29b883aeca8c67cd3043f2779692897e990650ff6d5216a3b6b27e0a6

SevKohler commented on src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/virologischerbefund/ProAnalytQuantitativesErgebnisElementConverter.java:
```diff
@@ -0,0 +1,26 @@
+
+        ProAnalytQuantitativesErgebnisElement proAnalytQuantitativesErgebnisElement = new ProAnalytQuantitativesErgebnisElement();
+        List <ProAnalytQuantitativesErgebnisChoice> proAnalytQuantitativesErgebnisChoiceList = new ArrayList<>();
+        if (observation.getValueQuantity().hasUnit()){
```
Unit is 1..1 no need for check if i see right.
So it will be always DvQuantity. 
Leave it be but maybe add an comment so people will now this is just an double check.

NormanWeinert replied:
Comment added.
https://github.com/ehrbase/fhir-bridge/pull/392/files#diff-3c21d1009f645927c78845f16c33a6c053c9c18394502f2ee4a12c28c9f28e80

SevKohler commented on src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/virologischerbefund/ProAnalytZugehoerigeLaborprobeChoiceConverter.java:
```diff
@@ -0,0 +1,56 @@
+
+        DvIdentifier dvIdentifier = new DvIdentifier();
+
+        if(observation.getSpecimen().getIdentifier() == null){
```
just put all of these in methods including the setters. 
So its a little bit more cleaned up. 

Identifier is 0..1 so just map it if its present.

Also not every field is required, therefore use Optional and only map what is existent. 
You have to check which fields are required for the DvIdentifier if some of them are present and others not, maybe set them with a blank field.

NormanWeinert replied:
Splitted up into mapping methods and boolean checker methods. Everything is also now an Optional.
https://github.com/ehrbase/fhir-bridge/pull/392/files#diff-585a1cf0f329e2ab64afa38a03d4bfe341376a4b2b2937848f1326f68da104f9

SevKohler commented on src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/virologischerbefund/ProAnalytQuantitativesErgebnisChoiceConverter.java:
```diff
@@ -0,0 +1,31 @@
+
+        ProAnalytQuantitativesErgebnisDvQuantity proAnalytQuantitativesErgebnisDvQuantity = new ProAnalytQuantitativesErgebnisDvQuantity();
+
+        proAnalytQuantitativesErgebnisDvQuantity.setQuantitativesErgebnisMagnitude(observation.getValueQuantity().getValue().doubleValue());
```
What happens if this is valuecodeable concept ? 
I cant see if this case is covered ?

NormanWeinert replied:
This was explained to me by the creator of the mapping:
In the case of ValueCodeableConcept only the Nachweistest is mapped after the code value. 
In the case of ValueQuantity Nachweistest is mapped to "Detected" and QuantitativesErgebnis will be mapped. QuantitativesErgebnis is an optional value.

SevKohler reviewed (changes requested):
Why are half of the fields from the template not mapped e.g. Interpretation, Externer Identifikator etc. please make sure every field is mapped, not just mandatory ones. 
There are equivalent representations for that in the fhir profles. 
When you added all mappings just request another review from me pls. 
Otherwise good job so far, its not an easy mapping!

For the fields that are mandatory in the template but not mandatory in the fhir resources (like the case of Fallidentifikation encounter is 0..1), you have to speak with the supervisors how to handle that, either we throw an exception like you did, or alter the template. 

There has been some major changes in the bridge due to clean ups and polishing, @subigre would you be so nice to explain him how to handle things in order to update his bridge to the changes.

subigre commented:
@SevKohler @NormanWeinert 

The problem comes from the fact that I try to fix the issue regarding the git repository by rewriting the history.
As a result this branch is now 1593 commits ahead of develop.

I have used two options in order to fix this issue with other branches:
1. Rebase the branch from develop and include changes at the end.
2. Create a patch including the changes, create a new branch from last develop and apply the patch.

NormanWeinert committed:
Refactoring after Code Review

NormanWeinert committed:
Revert "Reapply changes after rebase from develop"

This reverts commit aaf3f1b29efa590f20ccd7cb3e80b6226213d2a8.

NormanWeinert committed:
Revert "Update Profile.java"

This reverts commit 262d4a6a3572c466243ef1e7159b6259e9058f72.

NormanWeinert committed:
Revert "Revert "Reapply changes after rebase from develop""

This reverts commit aad759ce3b75bfc2c4032c59085fa137a5e76eb4.

NormanWeinert committed:
Revert "Reapply changes after rebase from develop"

This reverts commit aaf3f1b29efa590f20ccd7cb3e80b6226213d2a8.

NormanWeinert committed:
Revert "Revert "Update Profile.java""

This reverts commit 38f19f696b54f5b25de18acfb9e81f4ca022378c.

NormanWeinert committed:
Revert "Update Profile.java"

This reverts commit 262d4a6a3572c466243ef1e7159b6259e9058f72.

NormanWeinert committed:
Revert "Refactoring after Code Review"

This reverts commit 4c238fc1a374d1fbe56dd56a20ab0c2d7e830c16.

NormanWeinert committed:
Refactoring after Code Review

Addresses the issues mentioned in the code review.

SevKohler commented:
@NormanWeinert the CI is currently not working correctly which is fixed in #414

NormanWeinert commented:
@SevKohler  You cannot imagine the panic you lifted with that comment! I thought I broke something. Thanks!

SevKohler commented:
@NormanWeinert feel you haha, same goes for me i was checking stuff and recognised there is something wrong. Renaud then refered it was some new url coding we need to add, if not the CI build is runing into time outs.

SevKohler commented on src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/virologischerbefund/BefundJedesEreignisPointEventConverter.java:
```diff
@@ -0,0 +1,94 @@
+            mapZeitpunktDerProbenentnahme(specimen).ifPresent(probecluster::setZeitpunktDerProbenentnahmeValue);
+
+            AnatomischeLokalisationCluster anatomischeLokalisationCluster = new AnatomischeLokalisationCluster();
+            for (Coding loop : specimen.getCollection().getBodySite().getCoding()){
```
BodySite is 0..1 and Coding in BodySite too you have to unfortunatly check them too.

NormanWeinert replied:
Is now checked as well in an extra methods. "2 methods" as otherwise the if/for checks lead to a code smell.
https://github.com/ehrbase/fhir-bridge/pull/392/files#diff-7e8996e1193679795973b45c1d8e9b897e57859669d2d6a01138153dfe06ad49

SevKohler commented on src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/virologischerbefund/BefundJedesEreignisPointEventConverter.java:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    private boolean checkLabortestbezeichnung(Observation observation) {
+        for (Coding loop : observation.getCategory().get(0).getCoding()){
```
make another for loop for the Coding, please never use fixed identifiers. JSON files have no sorting order (Fhir maybe has but that does not garantee its order), so this is kind of a gamble.

NormanWeinert replied:
Made another loop, but needed to split the method up again due to a code smell being created if handled in only one method.
https://github.com/ehrbase/fhir-bridge/pull/392/files#diff-7e8996e1193679795973b45c1d8e9b897e57859669d2d6a01138153dfe06ad49

SevKohler commented on src/main/java/org/ehrbase/fhirbridge/ehr/converter/specific/virologischerbefund/ProAnalytZugehoerigeLaborprobeChoiceConverter.java:
```diff
@@ -0,0 +1,93 @@
+        mapId(observation).ifPresent(dvIdentifier::setId);
+        mapType(observation).ifPresent(dvIdentifier::setType);
+
+        return Optional.of(dvIdentifier);
```
i would suggest to add here a hasIdentifier and if not an Optional.empty(); is to be returned (does this make sense? since i dont see no check for identifier).
If identifier is there map it and return, if not dont map it and do nothing (also then an Optional.empty() should be returned by convertDvIdentifier since nothing is mapped).
Furthermore all methods here are public, im pretty sure most of them should be private at least.

If you dont have an empty output you also dont need optional :D in general

Nevertheless, we do have parser for that see [here](https://github.com/ehrbase/fhir-bridge/blob/develop/src/main/java/org/ehrbase/fhirbridge/ehr/converter/generic/DvIdentifierParser.java), this one has some errors as i can see it, maybe you take your stuff (since its better then mine) and put it there, i missed out some checks as it seems. Just dont forget the hasIdentifier check for Optional.empty()

NormanWeinert replied:
There is a hasIdentifier and an Optional.empty, but that is handled before we go into this class+method.
Here is the place where it happens (private Optional<ProAnalytZugehoerigeLaborprobeChoice> mapZugehoerigeLaborprobe  in ProAnalytClusterConverter):
https://github.com/ehrbase/fhir-bridge/pull/392/files#diff-4d8c593b51ad51f7cbd0a2addd7fb06a59da59dc9c48d4cf9a9e1edb3fa9dc5e

As I first need to determine which of the 2 methods (convertDvIdentifier and convertDvUri) needs to be used to map the resource the Identifier check is made earlier.
If there is an Identifier --> convertDvIdentifier
If there is no Identifier but a TypeElement --> convertDvUri
If there is no Identifier or TypeElement then Optional.empty() will be triggered.

Methods have been made private though:
https://github.com/ehrbase/fhir-bridge/pull/392/files#diff-585a1cf0f329e2ab64afa38a03d4bfe341376a4b2b2937848f1326f68da104f9

SevKohler commented:
@NormanWeinert good job! 
I got 3 more reviews for you then we should be good for the merge.
Please dont foreget to do what @subigre explained [here ](https://github.com/ehrbase/fhir-bridge/pull/392#issuecomment-886753896)after everything is fixed.
Option one is the easiest btw.

NormanWeinert commented:
@SevKohler I already did a "Rebase the branch from develop and include changes at the end." It was to no effect though and there are still 1000+ commits.

After everything is fixed I think I will create a new branch, add the changes by hand and create a new pull request.

SevKohler commented:
@subigre maybe you can help him out [here](https://github.com/ehrbase/fhir-bridge/pull/392#issuecomment-888357096)

SevKohler commented:
@NormanWeinert did you add the missing mappings ?

NormanWeinert commented:
@SevKohler Forgot to comment there: There are no "missing" mappings as only certain important mappings should be converted in the case of the Virologischer Befund. They have been marked for me by the creator of the mapping in a mapping file:
Status - DiagnosticReport.status
Fall-Kennung - DiagnosticReport.encounter (reference to Encounter)
Labortest-Bezeichnung - Virologischer Befund.category.coding.snomedct-subcategory
Name der Körperstelle - Specimen.collection.bodySite.coding.code
Zeitpunkt der Probenentnahme - specimen.collection.collected[x]
Virusnachweistest - Virologischer Befund.code
Nachweis - Virologischer Befund.value[x].valueCodeableConcept
Quantitatives Ergebnis - Virologischer Befund.value[x].valueQuantity
Ergebnis-Status - Virologischer Befund.status
Zugehörige Laborprobe - Virologischer Befund.specimen (reference to Specimen)

subigre commented:
I am not a Git expert, so please start by doing a backup of your changes. 
In addition, I have used this method only for the branches where I was able to identify the first commit with changes.
For the rest, I have created a fresh new branches and applied the changes using patchs.

In order to "remove" the previous commits, I used rebase with `--onto` parameter:
```
git rebase --onto develop <commit_id>
```
where `<commit_id>` is the first commit with changes for the specific feature branch.

NormanWeinert committed:
Insert Backup

NormanWeinert committed:
Manually apply "Updates code according to last version of openEHR_SDK dependency"

NormanWeinert committed:
Merge branch 'develop' into feature/NUM-128_Virologischer-Befund

NormanWeinert committed:
Fix SonarCloud Code smells

NormanWeinert commented:
@SevKohler @subigre  git rebase --onto develop <commit_id> did not do the trick.

I addressed the change requests.

I think let's just finish the review process and if everything is in order I will create a new "Merge/Feature/num 128 virologischer befund" branch and apply the changes there.
After I patched that branch to the state of this original branch here (without the large ammounts of commits of course) and all the tests ran through then it can finally be merged.

SevKohler commented:
@NormanWeinert okay seems good to me, about what fields are mapped you guys have to decide that. Just be sure the compositions generate are filled with everything :) After this review comments i think we are done, things look great so far.

NormanWeinert commented:
@SevKohler Ok, then I will create a new merge branch tomorrow, patch it up and open a new pull request with a hopefully fixed up history then.
